[retrigger] Add ability to search Embeds

XxInvictus opened the pull request:
- Add ability to search embeds in messages
- Normal embeds or recently posted URL embeds still cached work normally
- on_message_edit was added due to auto-generated URL preview embeds not cached being edited into the message AFTER the on_message is triggered.
- In order to process the above the logic is configured such that you require edits AND embeds enabled to search these generated embeds

XxInvictus committed:
[retrigger] Add ability to search Embeds

- Add ability to search embeds in messages
- Normal embeds or recently posted URL embeds still cached work normally
- on_message_edit was added due to auto-generated URL preview embeds not cached being edited into the message AFTER the on_message is triggered.
- In order to process the above the logic is configured such that you require edits AND embeds enabled to search these generated embeds

TrustyJAID commented on retrigger/triggerhandler.py:
```diff
@@ -276,6 +276,34 @@ async def on_raw_message_edit(self, payload: discord.RawMessageUpdateEvent) -> N
+            # If we can't find the channel ignore it
+            return
+        await self.check_triggers(after_message, True)
+
```
`on_raw_message_edit` already exists, this will cause duplicate triggers having it in twice.

XxInvictus replied:
I may have got something wrong in the logic here but on_raw_message_edit should only be triggering when it is `check_edits AND NOT embeds` meanwhile on_message_edit is `check_edits AND embeds`

TrustyJAID replied:
The problem with this is we're already fetching the message in `on_raw_message_edit` meaning every edited message is potentially pulling 2 of our 5/5s rate limit on message fetches which is not going to happen in this cog.  Here you just copied what already exists in `on_raw_message_edit` without even utilizing d.py internally already passing the new message to you in the event.

TrustyJAID replied:
I should clarify since I see what you mean now. The issue is your approach is not very DRY (Don't Repeat Yourself). We already have the logic for checking edits and we have a criteria that embeds only show up on an edited message so if both must be active we can utilize the already existing logic to get the message passed to the trigger checker. A trigger must have check edits enabled *and* read embeds in order to function so we're still filtering the edit fetch for any trigger that doesn't have check edits enabled, we don't need an extra check in there for embeds too. I would rather see a note that check edits is required for read embeds work properly than repeat the same code we have here since it is a requirement due to a discord limitation.

XxInvictus replied:
That's okay I do fully understand what you mean, it was entirely my fault for not digging deeper when I ran into an issue!

I originally tested with raw but couldn't get a successful output along with originally I was misreading the discord.py doco and thought it was doing something that was excluding embeds from the raw version. I have done some further digging and found the reason it didn't originally work was because I glazed over `if "content" not in payload.data`, not processing that raw only contains the edits in payload.data, not the full message. This meant that when an embed was edited it doesn't have a `content` it has `embeds` and as such was returning on the very first check.

I have a current working version locally which I can push up soon to fix this all up and bring it into the one :)

XxInvictus replied:
Removed extra listener and combined into `on_raw_message_edit`

TrustyJAID reviewed (changes requested):
This feature has been requested in the past and denied because users are unable to generate embeds and if they are that is a violation of Discord TOS via self botting. This cog does not scan any bot messages including webhook messages which can send embeds. If someone links something with an inappropriate embed or possibly inappropriate embed consider filtering the url instead.

XxInvictus commented:
So for a couple of examples:
- URL's are fluid and non-sensical, coupled with the variety of URL shortening services. If you know what the preview embed of an undesirable url is and can take then action on that "no matter what URL is used to repost it" (within reason). The age old example, Rick Roll Videos 😆 there are HUNDREDS of links for the same thing. This isn't foolproof, just an extra layer.

- You can very easily use embeds to bypass any filtering/triggers. Sure they can't create manual embeds but as you have mentioned URLs are still created. You don't even need to do anything fancy to bypass the triggers, just use Twitter, write a tweet, then link it in Discord and Discord will load the embed.

XxInvictus commented:
It is probably worth mentioning. Realistically and in my use case it was only for the purpose of being able to scan url embed contents, particularly Twitter embed contents. Any mention of "general" embed searching was just a side-affect of this.

TrustyJAID commented on retrigger/converters.py:
```diff
@@ -270,6 +270,7 @@ class Trigger:
         "role_mention",
         "everyone_mention",
         "nsfw",
+        "embeds",
```
```suggestion
        "read_embeds",
```

XxInvictus replied:
Updated

TrustyJAID commented on retrigger/converters.py:
```diff
@@ -312,6 +313,7 @@ def __init__(
         self.role_mention: bool = kwargs.get("role_mention", False)
         self.everyone_mention: bool = kwargs.get("everyone_mention", False)
         self.nsfw: bool = kwargs.get("nsfw", False)
+        self.embeds: bool = kwargs.get("embeds", False)
```
```suggestion
        self.read_embeds: bool = kwargs.get("read_embeds", False)
```

XxInvictus replied:
Updated

TrustyJAID commented on retrigger/converters.py:
```diff
@@ -463,6 +465,7 @@ async def to_json(self) -> dict:
             "everyone_mention": self.everyone_mention,
             "role_mention": self.role_mention,
             "nsfw": self.nsfw,
+            "embeds": self.embeds,
```
```suggestion
            "read_embeds": self.read_embeds,
```

XxInvictus replied:
Updated

TrustyJAID commented on retrigger/menus.py:
```diff
@@ -221,6 +221,8 @@ async def format_page(self, view: discord.ui.View, trigger: Trigger):
             info += _("__TTS__: **Enabled**\n")
         if trigger.chance:
             info += _("__Chance__: **1 in {number}**\n").format(number=trigger.chance)
+        if trigger.embeds:
```
```suggestion
        if trigger.read_embeds:
```

XxInvictus replied:
Updated

TrustyJAID commented on retrigger/retrigger.py:
```diff
@@ -771,6 +771,30 @@ async def toggle_nsfw(self, ctx: commands.Context, trigger: TriggerExists) -> No
+            return await self._no_trigger(ctx, trigger)
+        if not await self.can_edit(ctx.author, trigger):
+            return await self._not_authorized(ctx)
+        trigger.embeds = not trigger.embeds
```
```suggestion
        trigger.read_embeds = not trigger.read_embeds
```

XxInvictus replied:
Updated

TrustyJAID commented on retrigger/triggerhandler.py:
```diff
@@ -254,7 +254,7 @@ async def on_raw_message_edit(self, payload: discord.RawMessageUpdateEvent) -> N
         if await self.bot.cog_disabled_in_guild(self, guild):
             return
-        if not any(t.check_edits for t in self.triggers[guild.id].values()):
+        if not any(t.check_edits and not t.embeds for t in self.triggers[guild.id].values()):
```
```suggestion
        if not any(t.check_edits for t in self.triggers[guild.id].values()):
```

XxInvictus replied:
Updated after removing the extra on_message_edit

TrustyJAID commented on retrigger/triggerhandler.py:
```diff
@@ -368,6 +396,9 @@ async def check_triggers(self, message: discord.Message, edit: bool) -> None:
             if trigger.ocr_search and ALLOW_OCR:
                 content += await self.get_image_text(message)
 
+            if trigger.embeds and len(message.embeds) > 0:
```
```suggestion
            if trigger.read_embeds and len(message.embeds) > 0:
```

XxInvictus replied:
Updated

TrustyJAID commented on retrigger/triggerhandler.py:
```diff
@@ -368,6 +396,9 @@ async def check_triggers(self, message: discord.Message, edit: bool) -> None:
                 content += await self.get_image_text(message)
 
+            if trigger.embeds and len(message.embeds) > 0:
+                content += " " + " ".join(str(embed.to_dict()) for embed in message.embeds)
```
I am not a fan of just passing the string of the dictionary to the filter check. It can cause unnecessary problems for user to utilize this feature having to build their regex on embeds around a python dictionary object. It would be better to itemize out the important visible parts of the embed objects like title, description, url, etc. We can probably ignore things like the proxy_icon_url since those are discord cdn urls anyways and could give the false impression if someone has poor regex to filter discord links.

XxInvictus replied:
Sure thing, I will look at filtering this down, this was purely for the sake of not "over-complicating" it and adding a big chunk of logic. I will get some data on multiple different types of URL embeds and what they contain because I know that Discord treats things like Twitter and Youtube "special" adding specific fields for these platforms.

XxInvictus replied:
Okay, trying a few sources such as
- Twitter
- Youtube
- Github
- Reddit
- Generic URL

I think the below is what may be good to pick out:
- Author
  - Name
  - URL
- Title
- Description
- Type
- Provider
  - Name
- URL

Edit: took out image/video url, where they are relevant they will likely be in the Primary url or original message

TrustyJAID replied:
I don't believe  the `Type` is relevant, that's an internal discord embed type. It's only ever going to be one of [these.](https://discord.com/developers/docs/resources/channel#embed-object-embed-types) If you're filtering by any of those types of embeds you might as well just toggle the ability for users to embed links in the channel entirely.

You're also missing the `fields`, `footer`, `thumbnail`, and `image` sections of an embed which could contain content you may want filtered. A embed could link to one page but contain an image or thumbnail from another webpage you absolutely don't want shared for example.

 You don't need to investigate the different types of links to see this it's [documented here](https://discord.com/developers/docs/resources/channel#embed-object-embed-structurel) for all embed structures.

To keep the code clean I recommend building a method `def convert_embed_to_string(self, embed: discord.Embed) -> str` or something along those lines that pulls the necessary attributes of the embed into a structured string format that is not just the dictionary. I expect the end string would look something like 
```
Author: name
url
Title: title
Description: description
Provider: provider
URL: url
Field 1: field 1 name
field 1 value
Field 2: field 2 name
field 2 value
Footer: footer
```
In order to not complicate regex searches it may be even better to not provide the descriptions for each attribute and just provide the string content of each separated by a new line.

XxInvictus replied:
You are right that the fields are documented, but unsurprisingly Discord does not treat all links equally in how it fills the embed fields for different sites. That was the reason for testing a multitude of sites.

The `Type` isn't static for links, it isn't `link` for all links. In saying that you are right in that it is probably not needed, the only types I have seen in testing are `rich`, `video` and `link` all of which are pretty useless by themselves and there are probably other, better ways to filter these types of links.

To keep to the original scope of just covering auto-generated URL embeds below are some comments on fields based on my findings from testing with the various URL types above:
*Fields*
- Not used for nearly all URL embeds, only used for Twitter in which case it only contains the like/dislike count
*Footer*
- Again only used for Twitter, the only thing it contains is the Twitter logo and timestamp
*Thumbnail*
- Might actually be useful to filter on this, the thumbnail is an image from the page for "Generic URL's" for most other things it is pretty useless though. MIGHT provide a static image for Youtube but it is not clear how thumbnail is represented visually here.
*Image*
- I did remove this but you are right in that it should be added, in Twitter it contains an attached image URL and in Imgur tests it contains the image itself
*Video*
- Same as above, I removed it but with Imgur this contains the actual image URL even when a gallery link is used.

My view is that it will be best to leave the descriptions. This is because say for example you want to trigger on posts from a certain Twitter user that may have a very generic dictionary-based name. If you don't give the option to filter by `\bAuthorName: GenericUsername\b" then you could end up with it triggering on every single use of the word which matches the generic username.

That is only my view though, happy to remove them if you think it best :)

TrustyJAID replied:
It should be noted that while you've tested a handful of sites, and it looks like Discord does some custom backend work for things like twitter, technically all embed fields are possible to be manipulated through a variety of means. That is to say a user can host a webpage that contains all the necessary information to build a [completely custom](https://www.reddit.com/r/discordapp/comments/82p8i6/a_basic_tutorial_on_how_to_get_the_most_out_of/) embed so all fields should be accounted for. As mentioned the type is irrelevant for the searching we just want to search things that can be customized in some way. 

The string should also avoid having something like `Author: None` in situations where that embed field does not exist. I also want to point out that if I wanted to filter something like twitter from a specific user it would be trivial to just filter the twitter URL containing tweets from said user through already existing means.

XxInvictus replied:
That is interesting! I wasn't aware of that avenue of manipulating the Embed data, I thought we were accidentally creeping into bot or self-bot created custom embeds setting attributes. I will definitely expand that data set then 😄
I understand it would be trivial to filter Twitter user another way, that was just an easy example from the top field I was looking at without having to go too deep hahaha

XxInvictus replied:
New function has been created for removing unwanted embed fields and formatting everything as a sensible string. During testing I added a print, below is how a message containing a Twitter url embed looks afterwards:
```
redbot   | https://twitter.com/KaiyokoStar/status/1654427905907204097?s=20
redbot   | FooterText: Twitter
redbot   | FooterIcon_url: https://abs.twimg.com/icons/apple-touch-icon-192x192.png
redbot   | ImageUrl: https://pbs.twimg.com/media/FvW1ZDFWAAAh_N3.jpg
redbot   | AuthorName: Kaiyoko 💫 (@KaiyokoStar)
redbot   | AuthorIcon_url: https://pbs.twimg.com/profile_images/1634341028315496449/zRm5_IHw_400x400.jpg
redbot   | AuthorUrl: https://twitter.com/KaiyokoStar
redbot   | Fields: [{'name': 'Likes', 'value': '399', 'inline': True}]
redbot   | Description: Fashion Report Week 275 - Full Details #ffxiv #ff14 #ファッションチェック
redbot   | Url: https://twitter.com/KaiyokoStar/status/1654427905907204097?s=20
```

XxInvictus replied:
Got another revision still to go here to parse out the children of `Fields`

XxInvictus replied:
With that done here is an example with Embed field `Fields` parsed out. Given all fields are name-value rather than using Field1, Field2... I opted with appending the `name` value to Fields then using the `value` as the direct value:
```
redbot   | https://twitter.com/KaiyokoStar/status/1626493822220771328?s=20
redbot   | FooterText: Twitter
redbot   | FooterIcon_url: https://abs.twimg.com/icons/apple-touch-icon-192x192.png
redbot   | ImageUrl: https://pbs.twimg.com/media/FpJ3akhXwAU2hGD.jpg
redbot   | AuthorName: Kaiyoko 💫 (@KaiyokoStar)
redbot   | AuthorIcon_url: https://pbs.twimg.com/profile_images/1634341028315496449/zRm5_IHw_400x400.jpg
redbot   | AuthorUrl: https://twitter.com/KaiyokoStar
redbot   | FieldsLikes: 168
redbot   | Description: Fashion Report Week 264 - Full Details #ffxiv #ff14 #ファッションチェック
redbot   | Url: https://twitter.com/KaiyokoStar/status/1626493822220771328?s=20
```

TrustyJAID commented on retrigger/retrigger.py:
```diff
@@ -771,6 +771,30 @@ async def toggle_nsfw(self, ctx: commands.Context, trigger: TriggerExists) -> No
         msg = _("Trigger {name} NSFW set to: {nsfw}").format(name=trigger.name, nsfw=trigger.nsfw)
         await ctx.send(msg)
 
+    @_edit.command(name="embeds")
```
```suggestion
    @_edit.command(name="readembeds")
```

XxInvictus replied:
Updated

TrustyJAID commented on retrigger/retrigger.py:
```diff
@@ -771,6 +771,30 @@ async def toggle_nsfw(self, ctx: commands.Context, trigger: TriggerExists) -> No
 
+    @_edit.command(name="embeds")
+    @checks.mod_or_permissions(manage_messages=True)
+    async def toggle_embeds(self, ctx: commands.Context, trigger: TriggerExists) -> None:
```
```suggestion
    async def toggle_read_embeds(self, ctx: commands.Context, trigger: TriggerExists) -> None:
```

XxInvictus replied:
Updated

TrustyJAID reviewed (changes requested):
I want to also see the `embeds` changed to `read_embeds` similar to `read_filenames`. I don't want this to give the false impression that it will send an embed when it doesn't.

TrustyJAID commented on retrigger/README.md:
```diff
@@ -242,6 +242,9 @@ Edit various settings in a set trigger.
  - **readfilenames** Edit whether or not a trigger will append filenames of attachments to the search.
  - **deleteafter** Text triggers can have an optional delete_after time set, this can be used to edit it.
  - **chance** Triggers can be setup with a chance to occur in form of `1 in chance`.
+ - **embeds** Toggle whether the bot will search the contents of embeds, see note below for special consideration on URL preview embeds.
```
```suggestion
 - **readembeds** Toggle whether the bot will search the contents of embeds, see note below for special consideration on URL preview embeds.
```

XxInvictus replied:
Updated

XxInvictus committed:
Merge remote-tracking branch 'upstream/master'

# Conflicts:
#	retrigger/menus.py
#	retrigger/triggerhandler.py

XxInvictus committed:
Update Embed Option Naming and Rework Parsing of Embed Content

- Renamed the option to variants of Read Embeds to avoid confusion
- Removed superfluous on_message_edit listener
- Added parsing of embed content to remove any unwanted fields
- Properly format the embed contents into a sensible string structure

XxInvictus committed:
Remove Extra Print

- Removing extra print inserted for output string confirmation

XxInvictus commented on retrigger/triggerhandler.py:
```diff
@@ -254,10 +254,10 @@ async def on_raw_message_edit(self, payload: discord.RawMessageUpdateEvent) -> N
             # log.debug(f"No triggers in {guild=} have check_edits enabled")
             return
-        if "bot" in payload.data["author"]:
+        if "bot" in payload.data.get("author", []):
```
This is required as the URL auto-embed edit has no author field and errors on missing field.

XxInvictus committed:
Handle "Fields" of Embeds and Fix Missed embeds -> read_embeds

- Added handling of `Fields` field of embeds
- Removed `inline` from blacklist, due to Name/Value mapping of `Fields` it made more sense to direct reference them and ignore `inline`
- Fixed `trigger.embeds` not updated to `trigger.read_embeds` not picked up in previous testing due to trigger cache

TrustyJAID commented on retrigger/triggerhandler.py:
```diff
@@ -386,6 +388,24 @@ async def check_triggers(self, message: discord.Message, edit: bool) -> None:
+        for field, value in flattened_embed_dict.items():
+            embed_string += f"\n{field}: {value}"
+        return embed_string
+
```
```suggestion
    @staticmethod
    def convert_embed_to_string(embed: discord.Embed, embed_index: int = 0) -> str:
        embed_dict = embed.to_dict()
        flattened_embed_dict = {}
        field_blacklist = ["type", "color", "proxy_url", "height", "width", "proxy_icon_url"]
        for field, value in embed_dict.items():
            if field in field_blacklist:
                continue
            if isinstance(value, dict):
                for subfield in value:
                    if subfield in field_blacklist:
                        continue
                    flattened_embed_dict[f"{field.lower()}-{subfield.lower()}"] = value[subfield]
            elif isinstance(value, list):
                for field_index, embedfields in enumerate(value):
                    emfield_name = embedfields["name"].lower()
                    flattened_embed_dict[f"{field.lower()}-{field_index}-{emfield_name}"] = embedfields["value"]
            else:
                flattened_embed_dict[field.lower()] = value
        return "\n".join(f"embed-{embed_index}-{field}: {value}" for field, value in flattened_embed_dict.items())

```
Some minor changes here, I feel like the string should designate that it is an embed and what index the embed is at for things like twitter links that technically contain multiple embeds in order to display multiple images in the embed. 

This should also be a staticmethod since we don't need anything from its parent class passed into it. 

In order to improve regex parsing for end users making the names lowercase and separate extra info by `-` to make it easier to define what part of the embed a user may want to search. e.g. `embed-\d-footer-text:(.+)(?!\n)` would capture only the content of the embeds footer in group 1. 

I also reduced the number of iterations, the inline dictionary manipulation would technically be iterating over each dict twice when we can just ignore the blacklist during the first iteration.

I re-added the timestamp because I have some cool ideas for parsing the timestamp and potentially converting to a discord timestamp, not sure if anything will pan out from that but ISO timestamps look pretty straightforward to build regex for and could be useful.

This should also technically be added in `abc.py` as an abstract staticmethod but I can add that later.

TrustyJAID commented on retrigger/triggerhandler.py:
```diff
@@ -367,6 +367,8 @@ async def check_triggers(self, message: discord.Message, edit: bool) -> None:
             if trigger.ocr_search and ALLOW_OCR:
                 content += await self.get_image_text(message)
+            if trigger.read_embeds and len(message.embeds) > 0:
+                content += "\n".join(self.convert_embed_to_string(embed) for embed in message.embeds)
```
```suggestion
                content += "\n".join(self.convert_embed_to_string(embed, index) for index, embed in enumerate(message.embeds))
```
Required for changes suggested on `convert_embed_to_string`.

XxInvictus committed:
Apply suggestions from code review

Co-authored-by: TrustyJAID <TrustyJAID@gmail.com>

XxInvictus commented:
The suggestions make sense and look great, committed them in! :)

TrustyJAID commented:
Now just failing style checks. You can run `make reformat` on the repo and that should fix everything for you.

XxInvictus committed:
Makefile Reformat Cleanup

TrustyJAID reviewed (approved):
Thanks for being patient.

XxInvictus commented:
All good, this repo is your baby, I was happy to make changes to fit your standards and vision. I don't want to just come in and desecrate your repo 😉

TrustyJAID commented:
I should point out that in the future if you make a PR I import a logger that you can use instead of `print`. You can develop on Red with the `--debug` flag and then whenever you do `log.debug("some string")` it will print out the data to the console *and* save it to the bots logs files so you can go back and look later. The main benefit of this is if you have an extraneous `log.debug` regular users won't see it when they start their bot.
